domainservice/src/main/kotlin/com/example/delivery/repository/impl/RedisProductRepository.kt

```
@@ -0,0 +1,203 @@
+    val reactiveRedisTemplate: ReactiveRedisTemplate<String, ByteArray>,
+) : ProductRepository by mongoProductRepository {
+    private val mapper = jacksonObjectMapper()
+
+    @CircuitBreaker(name = "redisCircuitBreaker", fallbackMethod = "fallbackSave")
+    override fun save(product: MongoProduct): Mono<MongoProduct> {
+        return mongoProductRepository.save(product)
+            .flatMap {
+                reactiveRedisTemplate.opsForValue()
+                    .set(it.id.toString(), mapper.writeValueAsBytes(it), TIMEOUT_DURATION)
+                    .thenReturn(it)
+            }
+    }
+
+    @CircuitBreaker(name = "redisCircuitBreaker", fallbackMethod = "fallbackFindAllByIds")
+    override fun findAllByIds(productIds: List<String>): Flux<MongoProduct> {
+        return reactiveRedisTemplate.opsForValue()
+            .multiGet(productIds)
+            .map { list -> list.filterNotNull().map { mapper.readValue<MongoProduct>(it) } }
+            .flatMapMany { cachedProducts ->
+                val cachedProductIds = cachedProducts.map { it.id.toString() }
+                val nonCachedIds = productIds - cachedProductIds
+                if (nonCachedIds.isEmpty()) {
+                    cachedProducts.toFlux()
+                } else {
+                    mongoProductRepository.findAllByIds(nonCachedIds).mergeWith(cachedProducts.toFlux())
+                }
+            }
+    }
+
+    @CircuitBreaker(name = "redisCircuitBreaker", fallbackMethod = "fallbackFindById")
+    override fun findById(id: String): Mono<MongoProduct> {
+        return reactiveRedisTemplate.opsForValue()
+            .get(id)
+            .handle { product, sink ->
+                if (product.isEmpty()) {
+                    sink.complete()
+                } else {
+                    runCatching { mapper.readValue<MongoProduct>(product) }
+                        .onSuccess(sink::next)
+                        .onFailure(sink::error)
+                }
```

in case of empty product (value in Redis is empty byte array) you should throw your custom not found exception directly from sinl, because currently you are goin one more time to mongo